### PR TITLE
impl(generator): replace leading service comment with config

### PR DESCRIPTION
**Pull request by alevenberg**
Part of https://github.com/googleapis/google-cloud-cpp/issues/12873

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12982)

### CODE REVIEW

File: generator/internal/descriptor_utils.cc [hunk 1]
```diff
@@ -652,15 +652,37 @@ std::string GetEffectiveServiceName(VarsDictionary const& vars,
...
 }
 
+/// If a service name mapping exists, return the replacement comment.
+/// If it does not exist, return a null optional.
+/// Parses a command line argument in the form:
+/// {"service_name_to_comments": "service_a=comment_a,service_b=comment_b"}.
+absl::optional<std::string> GetReplacementComment(VarsDictionary const& vars,
+                                                  absl::string_view name) {
+  auto service_name_to_comments = vars.find("service_name_to_comments");
+  if (service_name_to_comments != vars.end()) {
```
Comment (coryan) [inline generator/internal/descriptor_utils.cc:662]: I would have short-circuit this with:

```suggestion
  if (service_name_to_comments == vars.end()) return absl::nullopt;
```

File: generator/internal/descriptor_utils.cc [hunk 2]
```diff
@@ -652,15 +652,37 @@ std::string GetEffectiveServiceName(VarsDictionary const& vars,
...
     std::vector<std::pair<std::string, std::string>> const& initial_values) {
   VarsDictionary vars(initial_values.begin(), initial_values.end());
   auto const& service_name = GetEffectiveServiceName(vars, descriptor.name());
   vars["product_options_page"] = OptionsGroup(vars["product_path"]);
   vars["additional_pb_header_paths"] = FormatAdditionalPbHeaderPaths(vars);
-  vars["class_comment_block"] =
-      FormatClassCommentsFromServiceComments(descriptor, service_name);
+  vars["class_comment_block"] = FormatClassCommentsFromServiceComments(
+      descriptor, service_name,
+      GetReplacementComment(vars, service_name).value_or(""));
```
Comment (coryan) [inline generator/internal/descriptor_utils.cc:685]: I am not sure if this suggestion improves things.  If you pass the `absl::optional` I think you can simplify the code below..

```suggestion
  vars["class_comment_block"] = FormatClassCommentsFromServiceComments(
      descriptor, service_name,
      GetReplacementComment(vars, service_name));
```

File: generator/internal/format_class_comments.cc [hunk 3]
```diff
@@ -54,15 +54,19 @@ auto constexpr kFixedClientComment = R"""(
...
+  // Use the service descriptor to populate the service_source_location.
   if (!service.GetSourceLocation(&service_source_location) ||
       service_source_location.leading_comments.empty()) {
     GCP_LOG(INFO) << __FILE__ << ":" << __LINE__ << ": " << service.full_name()
                   << " no leading_comments to format";
     formatted_comments = absl::StrCat(" ", service_name, "Client");
   } else {
+    if (!replacement_comment.empty()) {
+      service_source_location.leading_comments = replacement_comment;
+    }
```
Comment (coryan) [inline generator/internal/format_class_comments.cc:69]:
```suggestion
    service_source_location.leading_comments = replacement_comment.value_or(
        std::move(service_source_location.leading_comments));
```

File: generator/internal/descriptor_utils.cc [hunk 4]
```diff
@@ -652,15 +652,37 @@ std::string GetEffectiveServiceName(VarsDictionary const& vars,
...
+    std::map<absl::string_view, absl::string_view> map;
+    for (absl::string_view arg :
+         absl::StrSplit(service_name_to_comments->second, ',')) {
+      map.insert(absl::StrSplit(arg, absl::MaxSplits('=', 1)));
+    }
+    auto result = map.find(name);
+    if (result != map.end()) {
+      return std::string(result->second.data(), result->second.size());
+    }
+  }
```
Comment (coryan) [inline generator/internal/descriptor_utils.cc:672]: This is a $O(n ln(n) + ln(n))$ algorithm when you can do a $O(n)$

```suggestion
    for (absl::string_view arg :
         absl::StrSplit(service_name_to_comments->second, ',')) {
      std::pair<absl::string_view, absl::string_view> p = absl::StrSplit(arg, absl::MaxSplits('=', 1));
      if (p.first == name) return std::string(p.second.data(), p.second.size());
    }
    return absl::nullopt;
```

**Review by coryan — APPROVED** [review]
Just nits and suggestions of doubtful value.

File: generator/internal/descriptor_utils.cc [hunk 5]
```diff
@@ -652,15 +652,37 @@ std::string GetEffectiveServiceName(VarsDictionary const& vars,
   return std::string(name.data(), name.size());
 }
 
+/// If a service name mapping exists, return the replacement comment.
+/// If it does not exist, return a null optional.
+/// Parses a command line argument in the form:
+/// {"service_name_to_comments": "service_a=comment_a,service_b=comment_b"}.
+absl::optional<std::string> GetReplacementComment(VarsDictionary const& vars,
```
Comment (devbww) [inline generator/internal/descriptor_utils.cc:659]: If `GetReplacementComment()` and `GetEffectiveServiceName()` don't need to be declared in `descriptor_utils.h`, then they should be defined within the anonymous namespace, and nor do they deserve doxygen-level comments.

File: generator/internal/descriptor_utils_test.cc [hunk 6]
```diff
@@ -467,6 +467,141 @@ INSTANTIATE_TEST_SUITE_P(
       return std::get<0>(info.param);
     });
 
+class CreateServiceNameToCommentMappingTest : public CreateServiceVarsTest {};
```
Comment (devbww) [inline generator/internal/descriptor_utils_test.cc:470]: `using CreateServiceNameToCommentMappingTest = CreateServiceVarsTest` should suffice here.  No need for inheritence I believe.

File: generator/internal/format_class_comments.h [hunk 7]
```diff
@@ -22,12 +22,20 @@ namespace google {
 namespace cloud {
 namespace generator_internal {
 
-/// The function does not just use service.name() and takes @p service_name in
-/// case there exists a service_name_mapping argument that is renaming the
-/// service.
+/* The function formats class comments based on the pre-existing service
```
Comment (devbww) [inline generator/internal/format_class_comments.h:25]: Restore doxygen-ness?

File: generator/internal/descriptor_utils.cc (hunk 1 above)
Comment (alevenberg) [reply generator/internal/descriptor_utils.cc]: Done.

File: generator/internal/descriptor_utils.cc (hunk 4 above)
Comment (alevenberg) [reply generator/internal/descriptor_utils.cc]: DOne.

File: generator/internal/descriptor_utils.cc (hunk 2 above)
Comment (alevenberg) [reply generator/internal/descriptor_utils.cc]: Done

File: generator/internal/format_class_comments.cc (hunk 3 above)
Comment (alevenberg) [reply generator/internal/format_class_comments.cc]: Done

File: generator/internal/format_class_comments.h (hunk 7 above)
Comment (alevenberg) [reply generator/internal/format_class_comments.h]: I think adding the ** suffices.

Or do you mean make the whole thing a comment with ///

File: generator/internal/descriptor_utils.cc (hunk 4 above)
Comment (alevenberg) [reply generator/internal/descriptor_utils.cc]: Did for above as well.

File: generator/internal/descriptor_utils.cc (hunk 5 above)
Comment (alevenberg) [reply generator/internal/descriptor_utils.cc]: Moved to anon namespace and fixed the comments.

Didn't realize the `///` was a doxygen thing.

File: generator/internal/descriptor_utils_test.cc (hunk 6 above)
Comment (alevenberg) [reply generator/internal/descriptor_utils_test.cc]: Done for here and above.

File: generator/internal/format_class_comments.h [hunk 8]
```diff
@@ -22,12 +22,21 @@ namespace google {
 namespace cloud {
 namespace generator_internal {
 
-/// The function does not just use service.name() and takes @p service_name in
-/// case there exists a service_name_mapping argument that is renaming the
-/// service.
+/** The function formats class comments based on the pre-existing service
```
Comment (devbww) [inline generator/internal/format_class_comments.h:25]: Typically the comment text starts on the next line.

File: generator/internal/format_class_comments.h (hunk 7 above)
Comment (devbww) [reply generator/internal/format_class_comments.h]: Either way.

File: generator/internal/format_class_comments.h (hunk 8 above)
Comment (alevenberg) [reply generator/internal/format_class_comments.h:25]: Done 

Addressed in https://github.com/googleapis/google-cloud-cpp/pull/12983

File: generator/internal/descriptor_utils.cc [hunk 9]
```diff
@@ -555,6 +555,43 @@ std::string FormatAdditionalPbHeaderPaths(VarsDictionary& vars) {
...
 }
 
+// If a service name mapping exists, return the new name.
+// Parses a command line argument in the form:
+// {"service_name_mappings": "service_a=new_service_a,service=new_service"}.
+std::string GetEffectiveServiceName(VarsDictionary const& vars,
+                                    absl::string_view name) {
+  auto service_name_mappings = vars.find("service_name_mappings");
+  if (service_name_mappings == vars.end()) {
+    return std::string(name.data(), name.size());
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply generator/internal/descriptor_utils.cc:565]: Done

**Pull request merged**